src/OculusDriverNode.cpp

```diff
@@ -0,0 +1,186 @@
+}
+
+
+void OculusDriverNode::publishCurrentConfig(){
+    sonar_driver_interfaces::msg::SonarConfiguration configuration;
+
+    configuration.header = commonHeader_;
+
+    // Fill message with configuration data from sonar
+    configuration.fire_mode = sonar_->getFireMode();
+    configuration.frequency = sonar_->getOperatingFrequency();
+    configuration.ping_rate = sonar_->getPingRate();
+    configuration.beam_count = sonar_->getBeamCount();
+    configuration.beam_separation = sonar_->getBeamSeparation();
+    configuration.min_range = sonar_->getMinimumRange();
+    configuration.max_range = sonar_->getMaximumRange();
+    configuration.current_range = sonar_->getCurrentRange();
+    configuration.range_resolution = sonar_->getRangeResolution();
+    configuration.range_count = sonar_->getRangeBinCount();
+    configuration.horz_fov = sonar_->getHorzFOV();
+    configuration.vert_fov = sonar_->getVertFOV();
+    configuration.angular_resolution = sonar_->getAngularResolution();
+    configuration.gain = sonar_->getCurrentGain();
+    configuration.gain_assist = sonar_->gainAssistEnabled();
+    configuration.gamma = sonar_->getGamma();
+    configuration.speed_of_sound = sonar_->getSpeedOfSound();
+    configuration.salinity = sonar_->getSalinity();
+    configuration.temperature = sonar_->getTemperature();
+    configuration.pressure = sonar_->getPressure();
+    configuration.net_speed_limit = sonar_->getNetworkSpeedLimit();
+
+    // Publish the message on the OculusDriverNode
+    pub_configuration->publish(configuration);
+
+    // Publish the bearing table 
+    sonar_driver_interfaces::msg::SonarBearings msg_bearings;
+    msg_bearings.header = commonHeader_;
+    msg_bearings.bearings = sonar_->getBearingTable();
+    this->pub_bearings->publish(msg_bearings);
+    
+}
+
+
+void OculusDriverNode::cb_reconfiguration(const sonar_driver_interfaces::msg::SonarConfigurationChange::SharedPtr msg){
+    sonar_->configure(
+        msg->fire_mode,
+        msg->range,
+        msg->gain,
+        msg->speed_of_sound,
+        msg->salinity,
+        msg->gain_assist,
+        msg->gamma,
+        msg->net_speed_limit
+    );
+
+    sonar_->setPingRate(msg->ping_rate);
+}
+
+/// @brief Method to execute upon receival of a simplePingResult Message 
+/// @param sonar 
+/// @param image 
+void OculusDriverNode::cb_simplePingResult(std::unique_ptr<SonarImage>& image){
+    updateCommonHeader();
+    publishImage(image);
+    publishCurrentConfig();
+
+}
+
+
+void OculusDriverNode::updateCommonHeader(){
+    commonHeader_.frame_id = "sonar_0";
+    commonHeader_.stamp = this->get_clock()->now();
+}
+
+int main(int argc, char **argv){
+    rclcpp::init(argc, argv);
+    
+    std::shared_ptr<OculusDriverNode> node = std::make_shared<OculusDriverNode>("OculusDriverNode");
+
+    // Initialize and connect to sonar
+    node->sonar_->findAndConnect();            // This starts the thread that will process new images
```

Comment:
@EulersBoiler Please instantiate the sonar dependency separately and supply it into the node if needed via its constructor (dependency inversion).